### PR TITLE
Add deprecation for Calls,Events,Errors,Pallet

### PR DESCRIPTION
**Pull request by pkhry**
### Description: 
Adds support for `DeprecationStatus` info in `RuntimeMetadataIR` for Calls,Events,Errors and Pallet itself. 
depends on #4851.
####  Adds `deprecation_info` field to
- PalletCallMetadataIR
- PalletMetadataIR
- PalletEventMetadataIR
- PalletErrorMetadataIR
#### `deprecated_variants` for
- PalletEventMetadataIR
- PalletErrorMetadataIR
#### `deprecated_indexes` for
- PalletCallMetadataIR

There's also some test updates to make sure that deprecation attributes are getting propagated to the relevant structs.

see: [issue](https://github.com/paritytech/polkadot-sdk/issues/4098), Solution: A

### Notes 
The `deprecated_info` and `deprecated_variants/indexes` are done as separate fields for this mr. in next [PR](https://github.com/paritytech/polkadot-sdk/pull/4994) I refactored this as suggested into a single field.

### PS:
i used cargo +nightly fmt --all and cargo +nightly-2024-04-10 fmt --all and still getting line changes due to formatting, what's the correct incantation?

### CODE REVIEW

File: substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs [hunk 1]
```diff
@@ -220,9 +227,7 @@ fn expand_pallet_metadata_events(
 
 		quote! {
 			Some(
-				#scrate::__private::metadata_ir::PalletEventMetadataIR {
-					ty: #scrate::__private::scale_info::meta_type::<#pallet_event>()
-				}
+				#pallet_event::deprecation_info()
```
Comment (lexnv) [inline substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs:230]: nit: I think this can be called `event_metadata()`

File: substrate/primitives/metadata-ir/src/types.rs [hunk 2]
```diff
@@ -340,19 +343,29 @@ impl IntoPortable for StorageEntryTypeIR {
 pub struct PalletCallMetadataIR<T: Form = MetaForm> {
 	/// The corresponding enum type for the pallet call.
 	pub ty: T::Type,
+	/// Deprecation status of the pallet call itself
+	pub deprecation_info: DeprecationStatus<T>,
+	/// Deprecation status of the call indexes
+	pub deprecated_indexes: BTreeMap<usize, DeprecationStatus<T>>,
 }
```
Comment (lexnv) [inline substrate/primitives/metadata-ir/src/types.rs:350]: Could we group the deprecation infos? Some initial idea would be:

```rust
enum DeprecationMetadataIr {
    Fully(DeprecationStatusIr),
    Partially(BTreeMap<usize, DeprecationStatusIr<T>>),
}
```

I would also rename the `DeprecationStatus` to `DeprecationStatusIR` sometime, since we'll need to convert this to a frame_metadata::DeprecationStatus

File: substrate/primitives/metadata-ir/src/types.rs [hunk 3]
```diff
@@ -361,19 +374,29 @@ impl From<MetaType> for PalletCallMetadataIR {
 pub struct PalletEventMetadataIR<T: Form = MetaForm> {
 	/// The Event type.
 	pub ty: T::Type,
+	/// Deprecation status of the event itself
+	pub deprecation_info: DeprecationStatus<T>,
+	/// Deprecation status of the variants
+	pub deprecated_variants: BTreeMap<T::String, DeprecationStatus<T>>,
```
Comment (lexnv) [inline substrate/primitives/metadata-ir/src/types.rs:380]: nit: Would the parsing be more complex if we propagate here a variant index (`usize`) instead of variant name? Just to keep consistency with the calls

Comment (pkhry) [reply substrate/primitives/metadata-ir/src/types.rs:380]: hmm, i think it should be fine.

File: substrate/primitives/metadata-ir/src/types.rs (hunk 2 above)
Comment (pkhry) [reply substrate/primitives/metadata-ir/src/types.rs:350]: Can i do this after this MR, to apply the change to every other place of usage and updating the tests?

File: substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs (hunk 1 above)
Comment (pkhry) [reply substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs]: renamed

File: substrate/primitives/metadata-ir/src/types.rs (hunk 3 above)
Comment (pkhry) [reply substrate/primitives/metadata-ir/src/types.rs]: pushed the change with usize indexes instead of strings

File: substrate/primitives/metadata-ir/src/types.rs (hunk 2 above)
Comment (lexnv) [reply substrate/primitives/metadata-ir/src/types.rs:350]: Yep sounds like a plan!

File: substrate/frame/support/procedural/src/pallet/expand/event.rs [hunk 4]
```diff
@@ -170,5 +195,17 @@ pub fn expand_event(def: &mut Def) -> proc_macro2::TokenStream {
 		impl<#event_impl_gen> From<#event_ident<#event_use_gen>> for () #event_where_clause {
 			fn from(_: #event_ident<#event_use_gen>) {}
 		}
+
+		impl<#event_impl_gen> #event_ident<#event_use_gen> #event_where_clause {
+			#[allow(dead_code)]
+			#[doc(hidden)]
+			pub fn event_metadata<W: #frame_support::__private::scale_info::TypeInfo + 'static>() -> #frame_support::__private::metadata_ir::PalletEventMetadataIR {
```
Comment (thiolliere) [inline substrate/frame/support/procedural/src/pallet/expand/event.rs:202]: isn't `W` actually `Self`? why not bounding `TypeInfo` and `'static` on the impl directly?

Comment (pkhry) [reply substrate/frame/support/procedural/src/pallet/expand/event.rs:202]: I was hitting this issue in some cases [link](https://github.com/paritytech/polkadot-sdk/actions/runs/9836299016/job/27151718829?pr=4948#step:4:1987).
If there's a better solution for it, I'd like to use that instead, but honestly I have no idea how to do that

Comment (thiolliere) [reply substrate/frame/support/procedural/src/pallet/expand/event.rs:202]: Adding to the where clause: `Self: TypeInfo` should work:

```diff
diff --git a/substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs b/substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs
index 9ea7e2de620..27c295a7edf 100644
--- a/substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs
+++ b/substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs
@@ -225,7 +225,7 @@ fn expand_pallet_metadata_events(
 
                quote! {
                        Some(
-                               #pallet_event::event_metadata::<#pallet_event>()
+                               #pallet_event::event_metadata()
                        )
                }
        } else {
diff --git a/substrate/frame/support/procedural/src/pallet/expand/event.rs b/substrate/frame/support/procedural/src/pallet/expand/event.rs
index 947c5c83ff7..960001f8236 100644
--- a/substrate/frame/support/procedural/src/pallet/expand/event.rs
+++ b/substrate/frame/support/procedural/src/pallet/expand/event.rs
@@ -79,6 +79,10 @@ pub fn expand_event(def: &mut Def) -> proc_macro2::TokenStream {
                        unreachable!("Checked by event parser")
                }
        };
+       let event_where_clause_and_type_info = super::merge_where_clauses(&[
+               &event.where_clause,
+               &syn::parse_quote! { where Self: #frame_support::__private::scale_info::TypeInfo },
+       ]);
 
        // Phantom data is added for generic event.
        if event.gen_kind.is_generic() {
@@ -196,12 +200,14 @@ pub fn expand_event(def: &mut Def) -> proc_macro2::TokenStream {
                        fn from(_: #event_ident<#event_use_gen>) {}
                }
 
-               impl<#event_impl_gen> #event_ident<#event_use_gen> #event_where_clause {
+               impl<#event_impl_gen> #event_ident<#event_use_gen>
+                       #event_where_clause_and_type_info
+               {
                        #[allow(dead_code)]
                        #[doc(hidden)]
-                       pub fn event_metadata<W: #frame_support::__private::scale_info::TypeInfo + 'static>() -> #frame_support::__private::metadata_ir::PalletEventMetadataIR {
+                       pub fn event_metadata() -> #frame_support::__private::metadata_ir::PalletEventMetadataIR {
                                #frame_support::__private::metadata_ir::PalletEventMetadataIR {
-                                       ty: #frame_support::__private::scale_info::meta_type::<W>(),
+                                       ty: #frame_support::__private::scale_info::meta_type::<Self>(),
                                        deprecation_info: #deprecation_status,
                                        deprecated_variants: #variants
                                }
```

File: substrate/frame/support/procedural/src/pallet/expand/call.rs [hunk 5]
```diff
@@ -245,6 +245,33 @@ pub fn expand_call(def: &mut Def) -> proc_macro2::TokenStream {
 			}
 		});
 
+	let deprecation_status = if let Some(call) = def.call.as_ref() {
+		crate::deprecation::get_deprecation(&quote::quote! {#frame_support}, &call.attrs)
+			.expect("Correctly parse deprecation attributes")
```
Comment (thiolliere) [inline substrate/frame/support/procedural/src/pallet/expand/call.rs:250]: I don't think syn will actually enforce the correct parsing of `deprecated`. It would be better not to panic.

File: substrate/frame/support/procedural/src/pallet/expand/call.rs [hunk 6]
```diff
@@ -245,6 +245,33 @@ pub fn expand_call(def: &mut Def) -> proc_macro2::TokenStream {
...
+		quote::quote! { #frame_support::__private::metadata_ir::DeprecationStatus::NotDeprecated }
+			.to_string();
+
+	let indexes: Vec<proc_macro2::TokenStream> = methods
+		.iter()
+		.filter_map(|x| {
+			let key = x.call_index;
+			let deprecation_status =
+				crate::deprecation::get_deprecation(&quote::quote! {#frame_support}, &x.attrs)
+					.expect("Correctly parse deprecation attributes");
```
Comment (thiolliere) [inline substrate/frame/support/procedural/src/pallet/expand/call.rs:265]: same as above

File: substrate/frame/support/procedural/src/pallet/expand/call.rs [hunk 7]
```diff
@@ -245,6 +245,33 @@ pub fn expand_call(def: &mut Def) -> proc_macro2::TokenStream {
...
+			if deprecation_status.to_string() == default_deprecation_info {
```
Comment (thiolliere) [inline substrate/frame/support/procedural/src/pallet/expand/call.rs:266]: maybe not ideal to compare between string here, the TokenStream can be equivalent while the string is different.

File: substrate/frame/support/procedural/src/pallet/expand/pallet_struct.rs [hunk 8]
```diff
@@ -83,9 +83,7 @@ pub fn expand_pallet_struct(def: &mut Def) -> proc_macro2::TokenStream {
 			impl<#type_impl_gen> #pallet_ident<#type_use_gen> #config_where_clause {
 				#[doc(hidden)]
 				pub fn error_metadata() -> Option<#frame_support::__private::metadata_ir::PalletErrorMetadataIR> {
-					Some(#frame_support::__private::metadata_ir::PalletErrorMetadataIR {
-						ty: #frame_support::__private::scale_info::meta_type::<#error_ident<#type_use_gen>>()
-					})
+					Some(<#error_ident<#type_use_gen>>::error_metadata())
 				}
 			}
```
Comment (thiolliere) [inline substrate/frame/support/procedural/src/pallet/expand/pallet_struct.rs:88]: note that you could move this whole impl block inside `expand/error.rs` same as call

File: substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs [hunk 9]
```diff
@@ -58,7 +58,12 @@ pub fn expand_runtime_metadata(
 					#attr
 				}
 			});
+			let deprecation_info = {
```
Comment (lexnv) [inline substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs:61]: nit: We can move this below into an `exapnd_pallet_metadata_deprecation`

File: substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs [hunk 10]
```diff
@@ -220,9 +225,7 @@ fn expand_pallet_metadata_events(
 
 		quote! {
 			Some(
-				#scrate::__private::metadata_ir::PalletEventMetadataIR {
-					ty: #scrate::__private::scale_info::meta_type::<#pallet_event>()
-				}
+				#pallet_event::event_metadata::<#pallet_event>()
 			)
```
Comment (lexnv) [inline substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs:229]: Nice! This looks better!

File: substrate/frame/support/procedural/src/pallet/expand/call.rs (hunk 5 above)
Comment (lexnv) [reply substrate/frame/support/procedural/src/pallet/expand/call.rs:250]: Yep, that makes sense to me! We can turn these into `compile_err`

File: substrate/frame/support/procedural/src/runtime/parse/mod.rs [hunk 11]
```diff
@@ -120,7 +120,6 @@ pub struct ExplicitAllPalletsDeclaration {
 	pub name: Ident,
 	pub pallets: Vec<Pallet>,
 }
-
```
Comment (lexnv) [inline substrate/frame/support/procedural/src/runtime/parse/mod.rs:123]: nit: unrelated?

File: substrate/primitives/metadata-ir/src/types.rs [hunk 12]
```diff
@@ -138,6 +138,8 @@ pub struct PalletMetadataIR<T: Form = MetaForm> {
 	pub index: u8,
 	/// Pallet documentation.
 	pub docs: Vec<T::String>,
+	/// Deprecation info
+	pub deprecation_info: DeprecationStatus<T>,
```
Comment (lexnv) [inline substrate/primitives/metadata-ir/src/types.rs:142]: Thinking about the other PRs, what's the actual value this can hold for pallet metadata? It would only be possible to hold `FullyDeprecated` or `NotDeprecated`?

Comment (pkhry) [reply substrate/primitives/metadata-ir/src/types.rs:142]: NotDeprecated,deprecatedWithoutNote and Deprecated(note,since). I'm also a bit on the fence whether it makes sense to deprecate the whole pallet or not

**Comment by kianenigma** [comment]
Okay, this is good, but kind of big. Can you link the associated context and issue to the PR please?

File: substrate/frame/support/procedural/src/runtime/parse/mod.rs (hunk 11 above)
Comment (pkhry) [reply substrate/frame/support/procedural/src/runtime/parse/mod.rs]: should be removed now

File: substrate/frame/support/procedural/src/pallet/expand/call.rs (hunk 6 above)
Comment (pkhry) [reply substrate/frame/support/procedural/src/pallet/expand/call.rs]: done

File: substrate/frame/support/procedural/src/pallet/expand/call.rs (hunk 5 above)
Comment (pkhry) [reply substrate/frame/support/procedural/src/pallet/expand/call.rs]: done

File: substrate/primitives/metadata-ir/src/types.rs [hunk 13]
```diff
@@ -340,19 +343,29 @@ impl IntoPortable for StorageEntryTypeIR {
 pub struct PalletCallMetadataIR<T: Form = MetaForm> {
 	/// The corresponding enum type for the pallet call.
 	pub ty: T::Type,
+	/// Deprecation status of the pallet call itself
+	pub deprecation_info: DeprecationStatus<T>,
+	/// Deprecation status of the call indexes
+	pub deprecated_indexes: BTreeMap<u8, DeprecationStatus<T>>,
```
Comment (gui1117) [inline substrate/primitives/metadata-ir/src/types.rs:349]: We could use `Compact<u8>` instead of `u8` to be future proof. Same for the other enumeration types.

Comment (pkhry) [reply substrate/primitives/metadata-ir/src/types.rs:349]: pushed this change to #4994

File: substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs (hunk 9 above)
Comment (pkhry) [reply substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs]: done

File: substrate/frame/support/procedural/src/pallet/expand/call.rs [hunk 14]
```diff
@@ -245,6 +245,21 @@ pub fn expand_call(def: &mut Def) -> proc_macro2::TokenStream {
 			}
 		});
 
+	let deprecation = if let Some(call) = def.call.as_ref() {
```
Comment (lexnv) [inline substrate/frame/support/procedural/src/pallet/expand/call.rs:248]: nit: Maybe we can remove the if statement here.

Did not check this for compiling, similar to:
```rust
let attributes = def.call.as_ref().map(|call| call.attrs.as_ref()).unwrap_or(&[]);
// then we can call get_deprecated_enum once

crate::deprecation::get_deprecation_enum(
			&quote::quote! {#frame_support},
			attributes,
			methods.iter().map(|item| (item.call_index as u8, item.attrs.as_ref())),
		)
```

**Review by lexnv — APPROVED** [review]
Nice one, the code logic is good!  

I'm happy to see this merged into: https://github.com/paritytech/polkadot-sdk/pull/4851, then I'll have another look and see how it all fits together 👍

File: substrate/frame/support/procedural/src/pallet/expand/error.rs [hunk 15]
```diff
@@ -102,6 +102,17 @@ pub fn expand_error(def: &mut Def) -> proc_macro2::TokenStream {
...
 	let capture_docs = if cfg!(feature = "no-metadata-docs") { "never" } else { "always" };
 
+	let deprecation = crate::deprecation::get_deprecation_enum(
+		&quote::quote! {#frame_support},
+		&error.attrs,
+		error_item
+			.variants
+			.iter()
+			.enumerate()
+			.map(|(index, item)| (index as u8, item.attrs.as_ref())),
```
Comment (gui1117) [inline substrate/frame/support/procedural/src/pallet/expand/error.rs:112]: I don't know if we want to support it but:

User can currently write `#[codec(index = N)]` on variants of error and event. The pallet macro doesn't prevent it AFAIK.

If it happens the `index` here and codec index can be different.

note that scale-info reads the `#[codec(index = N)]` attributes to infer the index.

File: substrate/frame/support/procedural/src/pallet/expand/event.rs [hunk 16]
```diff
@@ -96,6 +95,17 @@ pub fn expand_event(def: &mut Def) -> proc_macro2::TokenStream {
...
 	}
 
+	let deprecation = crate::deprecation::get_deprecation_enum(
+		&quote::quote! {#frame_support},
+		&event.attrs,
+		event_item
+			.variants
+			.iter()
+			.enumerate()
+			.map(|(index, item)| (index as u8, item.attrs.as_ref())),
```
Comment (gui1117) [inline substrate/frame/support/procedural/src/pallet/expand/event.rs:105]: Same as error, some index can be given with `#[codec(index = N)]`

also interestingly such a code compiles, but I think it is not intended. (I don't think we need to support this. And macro should could return an error IMO)
```diff
diff --git a/substrate/frame/assets/src/lib.rs b/substrate/frame/assets/src/lib.rs
index 6dbce717a8e..f95d2ee2fc9 100644
--- a/substrate/frame/assets/src/lib.rs
+++ b/substrate/frame/assets/src/lib.rs
@@ -548,9 +548,10 @@ pub mod pallet {
 
        #[pallet::event]
        #[pallet::generate_deposit(pub(super) fn deposit_event)]
+       #[repr(u32)]
        pub enum Event<T: Config<I>, I: 'static = ()> {
                /// Some asset class was created.
-               Created { asset_id: T::AssetId, creator: T::AccountId, owner: T::AccountId },
+               Created { asset_id: T::AssetId, creator: T::AccountId, owner: T::AccountId } = 4,
                /// Some assets were issued.
                Issued { asset_id: T::AssetId, owner: T::AccountId, amount: T::Balance },
                /// Some assets were transferred.
```

**Review by gui1117 — COMMENTED** [review]
LGTM, except maybe we should handle the `#[codec(index = ..)]` attributes to get the correct index used by scale-info.

File: substrate/frame/support/procedural/src/pallet/expand/error.rs (hunk 15 above)
Comment (pkhry) [reply substrate/frame/support/procedural/src/pallet/expand/error.rs:112]: should we check the attribute of the variant then and if `codec.index` is not present then use the definition index?

File: substrate/frame/support/procedural/src/pallet/expand/event.rs (hunk 16 above)
Comment (gui1117) [reply substrate/frame/support/procedural/src/pallet/expand/event.rs:105]: I think user of metadata only have access to information from scale-info. So if it is written that variant N is deprecated they will match it with the variant number N from the point of view of the scale-info information.

So yes enum variant changes position when sent of the wire. Because the wire doesn't transmit the rust variant index.

That said, I also don't mind forbidding usage of `#[codec(index = N)]` for both error and event variants

Comment (pkhry) [reply substrate/frame/support/procedural/src/pallet/expand/event.rs]: Done, now when looking for `deprecated` attribute on `Event`'s or `Error`'s the code checks for discriminant or explicit `codec(index = N)` before using definition order.

File: substrate/frame/support/procedural/src/pallet/expand/error.rs (hunk 15 above)
Comment (pkhry) [reply substrate/frame/support/procedural/src/pallet/expand/error.rs]: pushed the updated code

File: substrate/frame/support/procedural/src/pallet/expand/call.rs (hunk 14 above)
Comment (pkhry) [reply substrate/frame/support/procedural/src/pallet/expand/call.rs]: done as suggested

File: substrate/frame/support/procedural/src/pallet/expand/error.rs (hunk 15 above)
Comment (gui1117) [reply substrate/frame/support/procedural/src/pallet/expand/error.rs]: Great, change looks good to me

File: substrate/frame/support/procedural/src/pallet/expand/event.rs (hunk 16 above)
Comment (gui1117) [reply substrate/frame/support/procedural/src/pallet/expand/event.rs]: Yes, looks good to me

File: substrate/frame/support/test/tests/enum_deprecation.rs [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+}
+impl SomeAssociation2 for u64 {
+	type _2 = u64;
+}
+
+#[frame_support::pallet]
+/// Pallet documentation
+// Comments should not be included in the pallet documentation
+#[pallet_doc("../../README.md")]
+#[doc = include_str!("../../README.md")]
```
Comment (lexnv) [inline substrate/frame/support/test/tests/enum_deprecation.rs:85]: nit: we can remove these and focus on the deprecation field

File: substrate/frame/support/test/tests/enum_deprecation.rs [hunk 18]
```diff
@@ -0,0 +1,352 @@
...
+	type BlockLength = ();
+	type DbWeight = ();
+	type Version = ();
+	type PalletInfo = PalletInfo;
+	type AccountData = ();
+	type OnNewAccount = ();
+	type OnKilledAccount = ();
+	type SystemWeightInfo = ();
+	type SS58Prefix = ();
+	type OnSetCode = ();
```
Comment (kianenigma) [inline substrate/frame/support/test/tests/enum_deprecation.rs:301]: All of this could have been removed because of `derive_impl`

This is one good reason to clean up all instances of `derive_impl` on `frame_system`, as evidently people do still copy paste it and are not fully familiar with it :)

**Comment by kianenigma** [comment]
So now if we mark any `Rust` types as `#[depreated]` this will be visible in the metadata, right? 

would be great if you make a quick note about this in https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/metadata/index.html. 

Moreover, is PAPI team aware of this? Perhaps a more public outreach in the Polkadot forum for the time being?
